### ISSUE TITLE
Seeking Advice: YOLOv8 Performance - fps CUDA ONNX 

### ISSUE THREAD

**Issue by LivingSparks10** [ultralytics/ultralytics#8602]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi there!

I've been experimenting with YOLOv8 by Ultralytics, and I'm perplexed about the performance I'm seeing. 

Despite trying various optimizations like using PyTorch, ONNX, and OpenVINO exported models, I'm still getting 35 frames per second for a 640x480 image. 

Here are my performance metrics on a notebook with an Nvidia GeForce RTX 3080 Ti GPU:
NO CUDA: 70ms - 13 fps
CUDA: 7ms - 35 fps
CUDA with image size 160x160: 35 fps
CUDA with 0 classes: 35 fps
CUDA with half precision (half True): 35 fps
CUDA with max detections set to 1: 35 fps
ONNX: 33 fps
OpenVINO: 32 fps

Surprisingly, my iPhone XS Max achieves 33 fps with the same model "yolov8n" (I've downloaded the YOLO iOS app).

I'm puzzled as to why the iPhone is comparable to my GPU, especially considering also that YOLOv5 with PyTorch yields better performance (37-38 fps)

I've tried different CUDA versions (11.8 and 12.1), but the results remain the same.

Here are some additional details:
torch.cuda.is_available() returns True
torch.cuda.get_device_name(0) indicates "NVIDIA GeForce RTX 3080 Ti Laptop GPU"

Am I overlooking something here?
I'd appreciate any insights or suggestions to improve the performance. Thank you in advance!

I share the minimum code I have used:
```
import cv2
from ultralytics import YOLO
import torch
import time
import numpy as np

# Load a pretrained YOLOv8n model
model = YOLO("yolov8n.pt")
# model = YOLO("yolov8n_orig.onnx", task="detect")
# model = YOLO("yolov8n_openvino_model/", task="detect")

def readCam(cameraIndex):
    # Used to store the last N frame times
    frame_times = [0] * 10  # Adjust the size of the array based on desired smoothing

    frame_idx = 0  # Index for the cyclic array
    new_frame_time = 0
    prev_frame_time = 0
    # Open the webcam
    cap = cv2.VideoCapture(cameraIndex)

    while True:
        # Capture frame-by-frame
        ret, frame = cap.read()

        if not ret:
            print("Error: Unable to capture frame")
            break

        # Run inference on the frame
        results = model(frame, device=0, verbose=False)  # list of Results objects
        # Visualize the results
        for i, r in enumerate(results):
            # Plot results image
            im_bgr = r.plot()  # BGR-order numpy array
            # Display the frame with detected objects

        # Record current frame time
        new_frame_time = time.time()

        fps = 1 / (new_frame_time - prev_frame_time)
        prev_frame_time = new_frame_time

        # Update the cyclic array
        frame_times[frame_idx] = fps
        frame_idx = (frame_idx + 1) % len(frame_times)  # Wrap around the array
        fps = np.mean(frame_times)

        # Convert and display FPS
        fps = f"{fps:.1f}"  # Format string with 2 decimal places
        cv2.putText(im_bgr, fps, (7, 70), cv2.FONT_HERSHEY_SIMPLEX, 3, (100, 255, 0), 3, cv2.LINE_AA)

        cv2.imshow("frame", im_bgr)

        # Exit if "q" is pressed
        if cv2.waitKey(1) & 0xFF == ord("q"):
            break

    # Release the capture
    cap.release()
    cv2.destroyAllWindows()


def get_video_capture_devices():
    devices = []
    for i in range(10):  # Assuming there are 10 possible devices
        cap = cv2.VideoCapture(i)
        if cap.isOpened():
            _, _ = cap.read()  # Read a frame to ensure the device is working
            devices.append((i, cap.get(cv2.CAP_PROP_FRAME_WIDTH), cap.get(cv2.CAP_PROP_FRAME_HEIGHT)))
            cap.release()
    return devices

if __name__ == "__main__":
    print(torch.cuda.is_available())
    print(torch.cuda.get_device_name(0))
    readCam(0)


```

**Comment by glenn-jocher** [comment]
@LivingSparks10 hey there! 🙌 It looks like you've done quite a bit of experimentation to optimize your YOLOv8 performance. Given the details you've shared, here are a couple of thoughts:

1. **TensorRT**: Since you're using an NVIDIA GPU, exporting your model to TensorRT and running inference with it could potentially boost your fps. TensorRT is specifically designed to maximize deep learning inference performance on NVIDIA GPUs. You can export your model to TensorRT format using:
   ```python
   model.export(format='engine')
   ```
   Then, load and run inference with the exported `.engine` model.

2. **Batch Size**: If you're processing video streams or multiple images simultaneously, increasing the batch size can improve throughput. This is because GPUs are highly parallel processors and can process multiple images more efficiently than one at a time.

3. **Model Complexity**: The `yolov8n` model is designed to be a balance between speed and accuracy. If fps is a critical factor for your application, consider using a smaller model variant if available, which can offer faster inference times at the cost of some accuracy.

4. **PyTorch Profiler**: To dive deeper into where the bottlenecks might be occurring, you can use the PyTorch Profiler. This tool can help you understand the performance of your model layer by layer, and identify any operations that are particularly time-consuming.

5. **CUDA Streams**: For advanced users, exploring CUDA streams can offer a way to parallelize data loading and model inference, potentially improving overall throughput.

It's indeed intriguing that the iPhone's performance is on par with the RTX 3080 Ti for this task. This could be due to the highly optimized nature of mobile inference engines and the specific model optimizations for mobile deployment.

I hope these suggestions help! Keep experimenting, and feel free to share any further questions or updates. 🚀

**Comment by LivingSparks10** [comment]
thanks, @glenn-jocher,
I've tried (after a tedious tensorRT installation on windows..🥲) but I gain like 2 fps (37fps)

Here's the detailed configuration I'm currently using:

```
Ultralytics YOLOv8 version 1.20
Python 3.10.6
torch 2.2.1+cu121
CUDA 12.1
Device: NVIDIA GeForce RTX 3080 Ti Laptop GPU
ONNX version 1.15.0 with opset 17
TensorRT version 9.0.1.post11.dev4
```

Regarding batch size, given the real-time streaming nature of the task, it seems challenging to parallelize the computation effectively.

Concerning model complexity, I've opted for the smallest model variant available directly from Ultralytics. However, I'm open to testing other models if you have recommendations.

 I'll explore optimizing performance further by leveraging tools like the PyTorch Profiler and Cuda Streams. 

Thanks for the suggestion!

**Comment by glenn-jocher** [comment]
Hey @LivingSparks10! 🌟

Great to hear you've made some progress with TensorRT, even if it's just a modest fps increase. Given your setup and the real-time nature of your application, here are a couple of quick thoughts:

- **Model Variant**: Since you're already using the smallest YOLOv8 variant, you're on the right track. Keep an eye on Ultralytics' updates, as they occasionally release even more efficient models.

- **PyTorch Profiler & CUDA Streams**: These tools can indeed provide deeper insights and potential performance boosts. It's a bit of a deep dive but can uncover some non-obvious optimizations.

- **Alternative Inference Engines**: Since you're on Windows, exploring ONNX Runtime might offer additional performance benefits, especially with ONNX models. It's generally straightforward to use and well-supported on Windows.

```python
import onnxruntime as ort

session = ort.InferenceSession("yolov8n.onnx")
outputs = session.run(None, {"input": your_input_here})
```

- **Efficiency in Real-Time Streaming**: For real-time tasks, ensuring minimal latency in each processing step (not just model inference) is crucial. Sometimes, pre-processing or post-processing can be the bottleneck.

Keep pushing the boundaries, and feel free to reach out if you hit any more snags or have more insights to share! 🚀

**Comment by rajhlinux** [comment]
> ### Search before asking
> * [x]  I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.
> 
> ### Question
> Hi there!
> 
> I've been experimenting with YOLOv8 by Ultralytics, and I'm perplexed about the performance I'm seeing.
> 
> Despite trying various optimizations like using PyTorch, ONNX, and OpenVINO exported models, I'm still getting 35 frames per second for a 640x480 image.
> 
> Here are my performance metrics on a notebook with an Nvidia GeForce RTX 3080 Ti GPU: NO CUDA: 70ms - 13 fps CUDA: 7ms - 35 fps CUDA with image size 160x160: 35 fps CUDA with 0 classes: 35 fps CUDA with half precision (half True): 35 fps CUDA with max detections set to 1: 35 fps ONNX: 33 fps OpenVINO: 32 fps
> 
> Surprisingly, my iPhone XS Max achieves 33 fps with the same model "yolov8n" (I've downloaded the YOLO iOS app).
> 
> I'm puzzled as to why the iPhone is comparable to my GPU, especially considering also that YOLOv5 with PyTorch yields better performance (37-38 fps)
> 
> I've tried different CUDA versions (11.8 and 12.1), but the results remain the same.
> 
> Here are some additional details: torch.cuda.is_available() returns True torch.cuda.get_device_name(0) indicates "NVIDIA GeForce RTX 3080 Ti Laptop GPU"
> 
> Am I overlooking something here? I'd appreciate any insights or suggestions to improve the performance. Thank you in advance!
> 
> I share the minimum code I have used:
> 
> ```
> import cv2
> from ultralytics import YOLO
> import torch
> import time
> import numpy as np
> 
> # Load a pretrained YOLOv8n model
> model = YOLO("yolov8n.pt")
> # model = YOLO("yolov8n_orig.onnx", task="detect")
> # model = YOLO("yolov8n_openvino_model/", task="detect")
> 
> def readCam(cameraIndex):
>     # Used to store the last N frame times
>     frame_times = [0] * 10  # Adjust the size of the array based on desired smoothing
> 
>     frame_idx = 0  # Index for the cyclic array
>     new_frame_time = 0
>     prev_frame_time = 0
>     # Open the webcam
>     cap = cv2.VideoCapture(cameraIndex)
> 
>     while True:
>         # Capture frame-by-frame
>         ret, frame = cap.read()
> 
>         if not ret:
>             print("Error: Unable to capture frame")
>             break
> 
>         # Run inference on the frame
>         results = model(frame, device=0, verbose=False)  # list of Results objects
>         # Visualize the results
>         for i, r in enumerate(results):
>             # Plot results image
>             im_bgr = r.plot()  # BGR-order numpy array
>             # Display the frame with detected objects
> 
>         # Record current frame time
>         new_frame_time = time.time()
> 
>         fps = 1 / (new_frame_time - prev_frame_time)
>         prev_frame_time = new_frame_time
> 
>         # Update the cyclic array
>         frame_times[frame_idx] = fps
>         frame_idx = (frame_idx + 1) % len(frame_times)  # Wrap around the array
>         fps = np.mean(frame_times)
> 
>         # Convert and display FPS
>         fps = f"{fps:.1f}"  # Format string with 2 decimal places
>         cv2.putText(im_bgr, fps, (7, 70), cv2.FONT_HERSHEY_SIMPLEX, 3, (100, 255, 0), 3, cv2.LINE_AA)
> 
>         cv2.imshow("frame", im_bgr)
> 
>         # Exit if "q" is pressed
>         if cv2.waitKey(1) & 0xFF == ord("q"):
>             break
> 
>     # Release the capture
>     cap.release()
>     cv2.destroyAllWindows()
> 
> 
> def get_video_capture_devices():
>     devices = []
>     for i in range(10):  # Assuming there are 10 possible devices
>         cap = cv2.VideoCapture(i)
>         if cap.isOpened():
>             _, _ = cap.read()  # Read a frame to ensure the device is working
>             devices.append((i, cap.get(cv2.CAP_PROP_FRAME_WIDTH), cap.get(cv2.CAP_PROP_FRAME_HEIGHT)))
>             cap.release()
>     return devices
> 
> if __name__ == "__main__":
>     print(torch.cuda.is_available())
>     print(torch.cuda.get_device_name(0))
>     readCam(0)
> ```
> 
> ### Additional
> _No response_

Here is some recommendations I can provide. I’m using the extra large Yolov8 model, extremely happy with the performance and results thus far. As for latency I can not confirm yet but it’s is real time, seems to be around the 20FPS at 640px. I’m using a 15 year old 8-core AMD CPU with a beefy RTX 3080 TI GPU. There will be bottleneck since I’m using such old CPU. My GPU is literally sleeping 😴 😂 during inference.

There are immense potential for improvements to gain FPS boosts. Such as:

Determining which codec input stream to use (might required to buy new camera hardware) and which processing hardware (CPU/GPU/PCIe Media cards) shall decode (or encode it if you plan to store the data).

Then the pre-preprocess and post-process all needs to be efficiently rewritten or use better logic (hardware and software wise) for max FPS, all needs to be determined which hardware will perform the computation.

btw I’m testing on ONNX GPU runtime via CUDA using uktralytics samples. I plan on to completely improve performance and then possibly start using TensorRT.

I will go as far as making custom FPGAs to make things faster if needed… this will be interesting.

**Comment by glenn-jocher** [comment]
Thank you for sharing detailed insights and your experimentation results. The performance you’re achieving aligns with what’s expected for the `yolov8n` model on your hardware, considering the balance between model complexity and frame rates. Optimizations on preprocessing, postprocessing, and hardware acceleration (e.g., TensorRT, FP16, or INT8 precision) can help, though significant FPS gains may require architectural or hardware-specific adjustments. For further guidance, you may reference our [TensorRT documentation](https://docs.ultralytics.com/integrations/tensorrt/) for advanced optimization techniques. Let us know if you have additional questions!
